lib/_emerge/EbuildBuild.py

```diff
@@ -165,8 +172,8 @@ def _prefetch_exit(self, prefetcher):
 				ebuild_path=self._ebuild_path,
 				fetchall=self.opts.fetch_all_uri,
 				fetchonly=self.opts.fetchonly,
-				background=False,
-				logfile=None,
+				background=quiet_setting,
+				logfile=fetch_log,
 				pkg=self.pkg,
 				scheduler=self.scheduler)
 			self._start_task(fetcher, self._fetchonly_exit)
```

Comment:
We should use `self.scheduler.fetch.schedule(fetcher)` to schedule this on the fetch queue, so that it has exclusive access to the fetch log. We'll need to somehow ensure that it's added to the queue here rather than started immediately, possibly by adding a method parameter that defaults to the status quo behavior:

https://github.com/gentoo/portage/blob/portage-3.0.18/lib/_emerge/Scheduler.py#L564

Reply:
I believe that my latest commit should add it to the scheduler queue without it starting immediately. I didn't end up altering the `_schedule_fetch()` method but hopefully I've been able to accomplish what you had in mind.

Reply:
I wondered if you were suggesting doing something like this with the new method param (superficially doesn't seem to behave differently from what I've committed but I might be wrong):

```
diff --git a/lib/_emerge/EbuildBuild.py b/lib/_emerge/EbuildBuild.py
index b71273dc6..58d2b9e4d 100644
--- a/lib/_emerge/EbuildBuild.py
+++ b/lib/_emerge/EbuildBuild.py
@@ -180,7 +180,7 @@ class EbuildBuild(CompositeTask):
 			if fetch_log:
 				fetcher.addExitListener(self._fetchonly_exit)
 				self._task_queued(fetcher)
-				self.scheduler.fetch.schedule(fetcher)
+				self.scheduler.fetch.schedule(fetcher, force_queue=True)
 			else:
 				self._start_task(fetcher, self._fetchonly_exit)
 			return
diff --git a/lib/_emerge/Scheduler.py b/lib/_emerge/Scheduler.py
index d3528ee68..3a9bef688 100644
--- a/lib/_emerge/Scheduler.py
+++ b/lib/_emerge/Scheduler.py
@@ -560,7 +560,7 @@ class Scheduler(PollScheduler):
 			if pargs:
 				self.status = pargs[0]
 
-	def _schedule_fetch(self, fetcher):
+	def _schedule_fetch(self, fetcher, force_queue=False):
 		"""
 		Schedule a fetcher, in order to control the number of concurrent
 		fetchers. If self._max_jobs is greater than 1 then the fetch
@@ -571,7 +571,7 @@ class Scheduler(PollScheduler):
 		would be required before it would be possible to enable
 		concurrent fetching within the parallel-fetch queue.
 		"""
-		if self._max_jobs > 1:
+		if self._max_jobs > 1 or not force_queue:
 			fetcher.start()
 		else:
 			self._task_queues.fetch.addFront(fetcher)
```

Reply:
The fetchonly flag triggers this `self._set_max_jobs(1)` call that happens to make it behave as desired:
```
lib/_emerge/Scheduler.py:383:   def _background_mode(self):
lib/_emerge/Scheduler.py:412:                                   self._set_max_jobs(1)
```
However, it's better if we go ahead and add a parameter to make this explicit. Your `force_queue` implementation loogs good, but the boolean expression needs an `and` like this:
```python
@@ -571,7 +571,7 @@ class Scheduler(PollScheduler):
 		would be required before it would be possible to enable
 		concurrent fetching within the parallel-fetch queue.
 		"""
-		if self._max_jobs > 1:
+		if self._max_jobs > 1 and not force_queue:
 			fetcher.start()
 		else:
 			self._task_queues.fetch.addFront(fetcher)
```

Reply:
That's the `force_queue` functionality added now. Thank you for taking the time to review my code! This was an excellent introduction to the portage codebase.

Reply:
You're welcome!